src/Microsoft.Identity.Web.DownstreamRestApi/DownstreamRestApi.cs

```diff
@@ -40,257 +41,124 @@ internal class DownstreamRestApi : IDownstreamRestApi
+        {
+            return CallRestApiInternalAsync(null, downstreamRestApiOptions, downstreamRestApiOptions.RequestAppToken, content,
+                                user, cancellationToken);
         }
 
         /// <inheritdoc/>
-        public async Task<HttpResponseMessage> CallRestApiForAppAsync(
-            string serviceName,
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Task<HttpResponseMessage> CallRestApiForUserAsync(
+            string? serviceName,
             Action<DownstreamRestApiOptions>? downstreamRestApiOptionsOverride = null,
+            ClaimsPrincipal? user = null,
             HttpContent? content = null,
             CancellationToken cancellationToken = default)
         {
             DownstreamRestApiOptions effectiveOptions = MergeOptions(serviceName, downstreamRestApiOptionsOverride);
+            return CallRestApiInternalAsync(serviceName, effectiveOptions, false, content, user, cancellationToken);
+        }
 
-            if (effectiveOptions.Scopes == null)
-            {
-                throw new ArgumentException(ScopesNotConfiguredInConfigurationOrViaDelegate);
-            }
-
-            string apiUrl = effectiveOptions.GetApiUrl();
-
-            using HttpRequestMessage httpRequestMessage = new HttpRequestMessage(
-                effectiveOptions.HttpMethod,
-                apiUrl);
-            if (content != null)
-            {
-                httpRequestMessage.Content = content;
-            }
-
-            string authorizationHeader = await
-                _authorizationHeaderProvider.CreateAuthorizationHeaderForAppAsync(
-                    effectiveOptions.Scopes.FirstOrDefault(),
-                    effectiveOptions,
-                    cancellationToken).ConfigureAwait(false);
-
-            httpRequestMessage.Headers.Add(Authorization, authorizationHeader);
-            effectiveOptions.CustomizeHttpRequestMessage?.Invoke(httpRequestMessage);
-            using HttpClient client = _httpClientFactory.CreateClient(serviceName);
-            return await client.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
+        /// <inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Task<HttpResponseMessage> CallRestApiForAppAsync(
+           string? serviceName,
+           Action<DownstreamRestApiOptions>? downstreamRestApiOptionsOverride = null,
+           HttpContent? content = null,
+           CancellationToken cancellationToken = default)
+        {
+            DownstreamRestApiOptions effectiveOptions = MergeOptions(serviceName, downstreamRestApiOptionsOverride);
+            return CallRestApiInternalAsync(serviceName, effectiveOptions, true, content, null, cancellationToken);
         }
 
         /// <inheritdoc/>
         public async Task<TOutput?> CallRestApiForUserAsync<TInput, TOutput>(
-            string serviceName,
+            string? serviceName,
```

Comment:
This one is marked async, but the ones above are not..also missing ` [MethodImpl(MethodImplOptions.AggressiveInlining)]`

Reply:
This one has 2 await, so it needs to be async. The others would have had an await just before the return.
Because it's async, not sure it can be inlined either.